sync/src/transaction_queue.rs

```diff
@@ -0,0 +1,616 @@
+		};
+
+		TxQueue {
+			current: current,
+			future: future,
+			last_nonces: HashMap::new(),
+			first_nonces: HashMap::new(),
+		}
+	}
+
+	/// Returns current status for this queue
+	pub fn status(&self) -> TxQueueStatus {
+		TxQueueStatus {
+			pending: self.current.priority.len(),
+			future: self.future.priority.len(),
+		}
+	}
+
+	/// Adds all signed transactions to queue to be verified and imported
+	pub fn add_all<T>(&mut self, txs: Vec<SignedTransaction>, fetch_nonce: T)
+		where T: Fn(&Address) -> U256 {
+		for tx in txs.into_iter() {
+			self.add(tx, &fetch_nonce);
+		}
+	}
+
+	/// Add signed transaction to queue to be verified and imported
+	pub fn add<T>(&mut self, tx: SignedTransaction, fetch_nonce: &T)
+		where T: Fn(&Address) -> U256 {
+		self.import_tx(tx, fetch_nonce);
+	}
+
+	/// Removes all transactions in given slice
+	///
+	/// If gap is introduced marks subsequent transactions as future
+	pub fn remove_all(&mut self, txs: &[SignedTransaction]) {
+		for tx in txs {
+			self.remove(&tx);
+		}
+	}
+
+	/// Removes transaction from queue.
+	///
+	/// If gap is introduced marks subsequent transactions as future
+	pub fn remove(&mut self, tx: &SignedTransaction) {
+		// Remove from current
+		let removed = self.current.remove(tx);
+		if let Some(verified_tx) = removed {
+			let sender = verified_tx.sender();
+
+			// Are there any other transactions from this sender?
+			if !self.current.address.has_row(&sender) {
+				// Clear last & first nonces
+				self.last_nonces.remove(&sender);
+				self.first_nonces.remove(&sender);
+				return;
+			}
+
+			// Let's find those with higher nonce (TODO optimize?)
+			let to_move_to_future = {
+				let row_map = self.current.address.get_row(&sender).unwrap();
+				let tx_nonce = verified_tx.tx.nonce;
+				let mut to_future = Vec::new();
+				let mut highest = U256::zero();
+				let mut lowest = tx_nonce.clone();
+
+				// Search nonces to remove and track lowest and highest
+				for (nonce, _) in row_map.iter() {
+					if nonce > &tx_nonce {
+						to_future.push(nonce.clone());
+					} else if nonce > &highest {
+						highest = nonce.clone();
+					} else if nonce < &lowest {
+						lowest = nonce.clone();
+					}
+				}
+
+				// Update first_nonces and last_nonces
+				if highest == U256::zero() {
+					self.last_nonces.remove(&sender);
+				} else {
+					self.last_nonces.insert(sender.clone(), highest);
+				}
+
+				if lowest == tx_nonce {
+					self.first_nonces.remove(&sender);
+				} else {
+					self.first_nonces.insert(sender.clone(), lowest);
+				}
+
+				// return to future
+				to_future
+			};
+
+			for k in to_move_to_future {
+				if let Some(v) = self.current.remove_by_address(&sender, &k) {
+					self.future.insert(sender.clone(), v.tx.nonce, v);
+				}
+			}
+			self.future.enforce_limit();
+			return;
+		}
+
+		// Remove from future
+		{
+			let sender = tx.sender().unwrap();
+			if let Some(_) = self.future.remove_by_address(&sender, &tx.nonce) {
```

Comment:
`if self.future.remove_by_address(&sender, &tx.nonce).is_some()` might be slightly lighter on syntax?